airbyte-cdk/python/airbyte_cdk/sources/declarative/manifest_declarative_source.py

```diff
@@ -46,18 +50,28 @@ class ManifestDeclarativeSource(DeclarativeSource):
 
     VALID_TOP_LEVEL_FIELDS = {"check", "definitions", "schemas", "spec", "streams", "type", "version"}
 
-    def __init__(self, source_config: ConnectionDefinition, debug: bool = False):
+    def __init__(self, source_config: ConnectionDefinition, debug: bool = False, construct_using_pydantic_models: bool = False):
```

Comment:
I'm wondering if it would have made sense to use a strategy pattern. My reasoning is that we see doing a migration generates a lot of change in this file. This is code that we will need to clean eventually and that will probably reappear in some for for every migration we will do. Having a strategy pattern allows us to not care about how the objects are created and therefore this class would stay the same whatever the migration is. 

So right now, one difference I see is that `ManifestDeclarativeSource` now pass as an argument the object it wants created. I think we could deduce it using the `type` in the pydantic model

Reply:
Sorry I might not be understanding exactly what you mean in how we might use the strategy pattern. This file should not undergo very many changes during the code its basically just the boolean to say use the old or new factory. Do you have a bit more of a tangible example to help me understand what you're suggesting better.

Once we finish this refactor in the next PR we would want to just remove the legacy factory in favor of the `model_to_component_factory.py` entirely so I'm not sure if we need to necessarily go the route of inserting this pattern for what should be short lived code.

Reply:
We would pass the strategy at the construction of the object like this:
```
class ManifestDeclarativeSource(DeclarativeSource):
    def __init__(self, source_config: ConnectionDefinition, debug: bool = False, models_construction_strategy: ModelConstructionStrategy):
        self.models_construction_strategy = models_construction_strategy
```

Now, everywhere where we need a model objects, we don't need the `if self.construct_using_pydantic_models:`. For example, `connection_checker` becomes:
```
    @property
    def connection_checker(self) -> ConnectionChecker:
        check = self._config["check"]
        if "type" not in check:
            check["type"] = "CheckStream"
        return self.models_construction_strategy.create_check_stream(CheckStreamModel, check, dict())(source=self)
```

This is interesting because:
* it allows `ManifestDeclarativeSource` to be decoupled from the implementation details. It doesn't have to care if it's pydantic, Django or even custom code. So when we have a new way of constructing objects, we can just pass the right strategy to `ManifestDeclarativeSource` and no need to add `if self.construct_using_pydantic_models` everywhere we construct objects. This means that the unit tests do not change. Sure we will probably add a couple of integration tests but those should be less
* it simplifies the migration. Instead of having to go back to all the `if self.construct_using_pydantic_models`, we can simply pass the other strategy to the constructor and that's it

Maybe this is overkill for here, I'm not so sure. I don't fully understand the whole context so don't hesitate to continue with the approach you were taking.

Reply:
Ah okay that makes sense, so each strategy adheres to a common interface, but depending on the strategy's method implementation that dictates which flow gets invoked. Thanks for clarifying.

Thinking about it a bit, if we were planning to support both flows or long term expect more component creation strategies I think this would be a good fit. But the plan is to roll over to the new flow in the other in flight PR so we'll end up with a single strategy very soon. Since just one strategy for one type might not be quite as beneficial, I think I'll leave this as is. But I'm gonna remember this idea in case we can use it in the future